algorithms/sycl/Reduction.cpp

```diff
@@ -32,57 +27,98 @@ inline size_t getNearestPow2Number(size_t number) {
   return (BASE << power);
 }
 
-template <typename T, typename OperationT> void reduce(size_t size, size_t step, T *buffer, OperationT operation, void* streamPtr) {
 
-  auto targetSize = getNearestPow2Number(size / (step * 2));
-  auto rng = ::device::internals::computeDefaultExecutionRange1D(targetSize);
+template <typename T, typename OperationT>
+void reduce(T *to, T *from, size_t reducedSize, OperationT Operation, void* queuePtr) {
+  auto rng = ::device::internals::computeDefaultExecutionRange1D(reducedSize);
+  auto queue = reinterpret_cast<cl::sycl::queue*>(queuePtr);
+
+  queue->submit([&](handler &cgh) {
+    auto numLanes = ::device::internals::WARP_SIZE;
+    cl::sycl::local_accessor<T> shrMem{numLanes, cgh};
 
-  ((cl::sycl::queue *) streamPtr)->submit([&](handler &cgh) {
     cgh.parallel_for(rng, [=](nd_item<> item) {
-      int j = item.get_global_id(0) * (2 * step);
-      if (j < size) {
-        T buddyValue = operation.getDefaultValue();
-        T currentValue = buffer[j];
-
-        int buddyIndex = j + step;
-        if (buddyIndex < size) {
-          buddyValue = buffer[buddyIndex];
+      auto localRange = item.get_local_range();
+      auto gid = item.get_global_id(0);
+      auto lid = item.get_local_id(0);
+
+      shrMem[lid] = from[gid];
+      item.barrier();
+
+      for (int offset = numLanes/2; offset >= 1; offset /= 2) {
+        if (lid < offset) {
+          shrMem[lid] = Operation(shrMem[lid], shrMem[lid + offset]);
         }
-        buffer[j] = operation(currentValue, buddyValue);
+        item.barrier();
+      }
+
+      auto wid = item.get_group().get_id(0);
+      if (lid == 0) {
+        to[wid] = shrMem[0];
       }
     });
   });
 }
 
-template <typename T> T Algorithms::reduceVector(T *buffer, size_t size, const ReductionType type, void* streamPtr) {
-  if (api == nullptr)
+
+template <typename T>
+T Algorithms::reduceVector(T *buffer, size_t size, const ReductionType type, void* queuePtr) {
+  if (api == nullptr) {
     throw std::invalid_argument("api has not been attached to algorithms sub-system");
+  }
+
+  size_t adjustedSize = getNearestPow2Number(size);
+  const size_t totalBuffersSize = 2 * adjustedSize * sizeof(T);
+
+  T *reductionBuffer = reinterpret_cast<T *>(api->getStackMemory(totalBuffersSize));
+
+  this->fillArray(reinterpret_cast<char *>(reductionBuffer),
+                  static_cast<char>(0),
+                  2 * adjustedSize * sizeof(T),
+                  queuePtr);
 
-  for (int step = 1; step < size; step *= 2) {
+  api->copyBetween(reductionBuffer, buffer, size * sizeof(T));
+
+  T *buffer0 = &reductionBuffer[0];
+  T *buffer1 = &reductionBuffer[adjustedSize];
```

Comment:
I'm not sure why, but this does look like an off-by-one error...

Reply:
`reductionBuffer` was allocated with size `2 * adjustedSize`. Addresses of the first half span from `0` to `adjustedSize - 1`. The second half spans between `adjustedSize` and `2 * adjustedSize - 1`.

In general, I allocate, copy and preform a reduction in a new buffer because I don't wanna corrupt the user's buffer because the user maybe need the original one for something else.

Of course, there are some vendor specific libraries which have collections of reduction algorithms. However, you do not know in which way they allocate auxiliary memory. In our case, we can take an advantage of the stack memory and avoid latency of GPU memory allocation. In my original implementation, which I did a year ago, I used `Nvidia thrust` but then I decide to switch to my own implementation